packages/flutter_tools/lib/src/commands/clean.dart

```diff
@@ -28,38 +32,71 @@ class CleanCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    // Clean Xcode to remove intermediate DerivedData artifacts.
+    // Do this before removing ephemeral directory, which would delete the xcworkspace.
+    final FlutterProject flutterProject = FlutterProject.current();
+    if (xcode.isInstalledAndMeetsVersionCheck) {
+      await _cleanXcode(flutterProject.ios);
+      await _cleanXcode(flutterProject.macos);
+    }
+
     final Directory buildDir = fs.directory(getBuildDirectory());
-    _deleteFile(buildDir);
+    deleteFile(buildDir);
 
-    final FlutterProject flutterProject = FlutterProject.current();
-    _deleteFile(flutterProject.dartTool);
+    deleteFile(flutterProject.dartTool);
 
     final Directory androidEphemeralDirectory = flutterProject.android.ephemeralDirectory;
-    _deleteFile(androidEphemeralDirectory);
+    deleteFile(androidEphemeralDirectory);
 
     final Directory iosEphemeralDirectory = flutterProject.ios.ephemeralDirectory;
-    _deleteFile(iosEphemeralDirectory);
+    deleteFile(iosEphemeralDirectory);
+
+    final Directory macosEphemeralDirectory = flutterProject.macos.ephemeralDirectory;
+    deleteFile(macosEphemeralDirectory);
 
     return const FlutterCommandResult(ExitStatus.success);
   }
 
-  void _deleteFile(FileSystemEntity file) {
-    final String path = file.path;
-    printStatus("Deleting '$path${fs.path.separator}'.");
-    if (file.existsSync()) {
-      try {
-        file.deleteSync(recursive: true);
-      } on FileSystemException catch (error) {
-        if (platform.isWindows) {
-          printError(
-            'Failed to remove $path. '
+  Future<void> _cleanXcode(XcodeBasedProject xcodeProject) async {
+    if (!xcodeProject.existsSync()) {
+      return;
+    }
+    Status xcodeStatus;
+    try {
+      final Directory xcodeWorkspace = xcodeProject.xcodeWorkspace;
+      final XcodeProjectInfo projectInfo = await xcodeProjectInterpreter.getInfo(xcodeWorkspace.parent.path);
+      for (String scheme in projectInfo.schemes) {
+        xcodeStatus = logger.startProgress('Cleaning Xcode workspace for scheme $scheme...', timeout: timeoutConfiguration.slowOperation);
+        xcodeProjectInterpreter.cleanWorkspace(xcodeWorkspace.path, scheme);
+      }
+    } catch (error) {
+      printTrace('Could not clean Xcode workspace: $error');
+    } finally {
+      xcodeStatus?.stop();
+    }
+  }
+
+  @visibleForTesting
+  void deleteFile(FileSystemEntity file) {
+    if (!file.existsSync()) {
+      return;
+    }
+    final Status deletionStatus = logger.startProgress('Deleting ${file.basename}...', timeout: timeoutConfiguration.fastOperation);
```

Comment:
It seemed surprising in use that this showed the progress spinner; it was very fast.